### ISSUE TITLE
RFC: Restructure ImageDecoder trait

### ISSUE THREAD

**Issue by fintelia** [PistonDevelopers/image#798]
I propose we break the `ImageDecoder` trait into several different traits targeting particular uses. These new traits consume self where appropriate to accommodate readers that can't seek, and `read_scanline` has been replaced with `into_box_read` and `read_image_with_progress`. I also introduce the concept of scanline_height so that decoders can express the minimum meaningful amount of data that they can decode without requiring internal buffering. Additionally, the new trait assumes that the image header has already been parsed, which means that metadata getter functions no longer need &mut self and cannot fail.

```rust
// This trait should be implementable for all image types, and should not require 
// `R: Seek` for any image types that don't need that bound for normal operation.
trait ImageDecoder {
    fn dimensions(&self) -> (u32, u32);
    fn colortype(&self) -> ColorType;
    fn row_bytes(&self) -> usize;
    fn read_image(self) -> ImageResult<DecodingResult>;

    // Returns minimum number of bytes that must be decoded at once to avoid buffering.
    // This could be as few as one byte or as many as the entire image.
    fn scanline_bytes(&self) -> usize;

    // Returns a reader that is intended to be called with slices of some multiple of 
    // scanline_bytes, but decodes into an internal buffer when required to read into
    // shorter slices.
    fn into_box_read(self) -> Box<Read>;
}

// This trait requires an underlying reader that has the ability to `Seek`.
trait ImageDecoderExt: ImageDecoder {
    // If possible for the image type, this function uses seek() to load only the
    // needed parts of the image into memory. Regardless, it then seeks back to the start
    // so `read_image` and friends still work.
    fn read_rect(&mut self, x: u32, y: u32, width: u32, height: u32, buf: &mut [u8]) -> ImageResult<()>;
}

// Non-object safe trait that lets users observe progress of image loading.
trait ImageDecoderIncremental {
    fn read_image_with_progress<F: Fn(f64)>(self, progress_callback: F) -> ImageResult<DecodingResult>;
}

// Only implemented for image types that support animation.
trait AnimationDecoder {
    fn into_frames(self) -> ImageResult<Frames>;
}
```

A typical decoder would then be implemented something like:
```rust
struct MyDecoder<R: Read> {}
impl<R: Read> ImageDecoder for MyDecoder<R> {}
impl<R: Read> ImageDecoderProgress for MyDecoder<R> {}
impl<R: Read + Seek> ImageDecoderExt for MyDecoder<R> {}
```

**Comment by bvssvni** [comment]
Looks good to me.

**Comment by ArturKovacs** [comment]
As suggested in #823 the `ImageDecoder` or `AnimationDecoder` trait's `into_frames` function could return an iterator that instead of storing all frames in memory, reads a single frame on the `next` function call. Such an iterator would allow access to a frame before the remaining frames are loaded.

@fintelia if you'd be interested in implementing this, it would certainly be nice to have this done in the associated pull request (#819) but otherwise I would also be happy to work on it. What would you suggest?

**Comment by fintelia** [comment]
@ArturKovacs I think that would be worth implementing, but I don't have the time to commit to it at the moment. Not sure the best way to go about coordinating; perhaps we could both push our work to a new branch and then merge into master once everything comes together? While we're making a breaking change, there also might be other parts of the API that should also be cleanup up...

**Comment by fintelia** [comment]
This is now merged via #819
